sns/parity_math.c

```diff
@@ -32,50 +32,195 @@
+ *                              to encode or decode data.
+ * @param[out] g_tbls         - Pointer to concatenated output tables for decode
+ * @retval     0              - success otherwise failure
+ */
+static int isal_gen_recov_coeff_tbl(uint32_t data_count, uint32_t parity_count,
+				    struct m0_buf *failed_idx_buf,
+				    struct m0_buf *alive_idx_buf,
+				    uint8_t *encode_mat, uint8_t *g_tbls);
+#endif /* ISAL_ENCODE_ENABLED */
+
+#if ISAL_ENCODE_ENABLED
+/**
+ * Sort the data and parity buffers based on input fail buffer. If buffer is
+ * marked as failed, its pointer will be added in frags_out buffer array. If
+ * buffer is not marked as failed in fail buffer, its pointer will be added in
+ * frags_in buffer array. Buffer array frags_in will be used as source buffers
+ * for recovery. Buffer array frags_out will be used as buffers to be recovered.
+ * @param[out] frags_in   - Array of buffer pointers containing pointers of
+ *                          buffers which are not failed.
+ * @param[out] frags_out  - Array of buffer pointers containing pointers of
+ *                          failed buffers.
+ * @param[in]  unit_count - Total count of buffers i.e. data_count + parity_count
+ * @param[in]  data_count - count of SNS data units used in system.
+ * @param[in]  fail       - block with flags, treated as uint8_t block with
+ *                          b_nob elements, if element is '1' then data or parity
+ *                          block with given index is treated as broken.
+ * @param[in]  data       - data block, treated as uint8_t block with
+ *                          b_nob elements.
+ * @param[in]  parity     - parity block, treated as uint8_t block with
+ *                          b_nob elements.
+ * @retval     true         on success
+ * @retval     false        on failure to sort buffers
+ */
+static bool buf_sort(uint8_t **frags_in, uint8_t **frags_out,
+		    uint32_t unit_count, uint32_t data_count,
+		    uint8_t *fail, struct m0_buf *data,
+		    struct m0_buf *parity);
+#endif /* ISAL_ENCODE_ENABLED */
+
+#if ISAL_ENCODE_ENABLED
+/**
+ * Sort the indices for failed and non-failed data and parity blocks.
+ * @param[in]  fail       - block with flags, if element is '1' then data or parity
+ *                          block with given index is treated as broken.
+ * @param[in]  unit_count - Total length of fail buffer
+ * @param[out] failed_idx - block with failed indices, treated as uint8_t block
+ *                          with b_nob elements
+ * @param[out] alive_idx  - block with non-failed (alive) indices, treated as
+ *                          uint8_t block with b_nob elements
+ * @retval     true         on success
+ * @retval     false        on failure to sort indices
+ */
+static bool fails_sort(uint8_t *fail, uint32_t unit_count,
+		       struct m0_buf *failed_idx, struct m0_buf *alive_idx);
+#endif /* ISAL_ENCODE_ENABLED */
```

Comment:
From line 90 to 185, all can be under single Macro of ISAL enabled.

Reply:
Done.